_posts/2017-02-22-ChainerRL-Deep-Reinforcement-Learning-Library.md

```diff
@@ -0,0 +1,114 @@
+
+{% highlight python linenos %}
+env = YourEnv()
+# reset() returns the current observation given the envirnment
+obs = env.reset()
+action = 0
+# step() sends an action to the environemnt, then returns four values (next observation, reward, whether it reachs the terminal of episode, and additional information).
+obs, r, done, info = env.step(action)
+{% endhighlight %}
+
+In DRL, neural networks correspond to policy and value functions (V-function or Q-function), which determines action given state, and estimates value of state or action, respectively. The parameters of neural network models are then updated through training. In ChainerRL, policies and value functions are represented as `Link` of Chainer, which just implements `__call__()` method.
+
+{% highlight python linenos %}
+class CustomDiscreteQFunction(chainer.Chain):
+    def __init__(self):
+        super().__init__(l1=L.Linear(100, 50)
+                         l2=L.Linear(50, 4))
+    def __call__(self, x, test=False):
+        h = F.relu(self.l1(x))
+        h = self.l2(h)
+        return chainerrl.action_value.DiscreteActionValue(h)
+ 
+class CustomGaussianPolicy(chainer.Chain):
+    def __init__(self):
+        super().__init__(l1=L.Linear(100, 50)
+                         mean=L.Linear(50, 4),
+                         var=L.Linear(50, 4))
+    def __call__(self, x, test=False):
+        h = F.relu(self.l1(x))
+        mean = self.mean(h)
+        var = self.var(h)
+        return chainerrl.distribution.GaussianDistribution(mean, var)
+{% endhighlight %}
+
+Then "Agent" can be defined given the model, an optimizer in Chainer, and algorithm-specific parameters. Agents execute the training of the model through interactions with the environment.
+
+{% highlight python linenos %}
+q_func = CustomDiscreteQFunction()
+optimizer = chainer.Adam()
+optimizer.setup(q_func)
+agent = chainerrl.agents.DQN(q_func, optimizer, ...)  # truncated other parameters
+{% endhighlight %}
+
+After creating the agent, training can be done either by user's own training loop,
+
+{%highlight python linenos %}
+# Training
+obs = env.reset()
+r = 0
+done = False
+for _ in range(10000):
+    while not done:
+        action = agent.act_and_train(obs, r)
+        obs, r, done, info = env.step(action)
+    agent.stop_episode_and_train(obs, r, done)
+    obs = env.reset()
+    r = 0
+    done = False
+agent.save('final_agent')
+{% endhighlight %}
+
+or a pre-defined training function as follows.
+
+{% highlight python linenos %}
+chainerrl.experiments.train_agent_with_evaluation(
+    agent, env, steps=100000, eval_frequency=10000, eval_n_runs=10,
+    outdir='results')
+{% endhighlight %}
+
+We also provide a quickstart guide to start playing with ChainerRL.
+[https://github.com/pfnet/chainerrl/blob/master/examples/quickstart/quickstart.ipynb](https://github.com/pfnet/chainerrl/blob/master/examples/quickstart/quickstart.ipynb)
```

Comment:
The full URL makes the document difficult to read (at least to me). How about making a link to it with `[quickstart guide](...)`, although the original post does not do so.